tools/make_patch.c

```diff
@@ -0,0 +1,437 @@
+#define PROGRAM_NAME "make_patch"
+#define USAGE_OPTS "labels.sym constants.sym patched.gbc original.gbc vc.patch.template vc.patch"
+
+#include "common.h"
+
+#include <ctype.h>
+
+struct Buffer {
+	size_t item_size;
+	size_t size;
+	size_t capacity;
+	void *data;
+};
+
+struct Symbol {
+	struct Symbol *next;
+	unsigned int address;
+	unsigned int offset;
+	char name[]; // C99 FAM
+};
+
+struct Patch {
+	unsigned int offset;
+	unsigned int size;
+};
+
+struct Buffer *buffer_create(size_t item_size) {
+	struct Buffer *buffer = xmalloc(sizeof(*buffer));
+	buffer->item_size = item_size;
+	buffer->size = 0;
+	buffer->capacity = 0x10;
+	buffer->data = xmalloc(buffer->capacity * item_size);
+	return buffer;
+}
+
+void buffer_append(struct Buffer *buffer, const void *item) {
+	if (buffer->size >= buffer->capacity) {
+		buffer->capacity = (buffer->capacity + 1) * 2;
+		buffer->data = xrealloc(buffer->data, buffer->capacity * buffer->item_size);
+	}
+	memcpy((char *)buffer->data + (buffer->size++ * buffer->item_size), item, buffer->item_size);
+}
+
+void buffer_free(struct Buffer *buffer) {
+	free(buffer->data);
+	free(buffer);
+}
+
+int get_address_type_limit(int address) {
+	return address < 0x8000 ? 0x4000 // ROM
+	     : address < 0xa000 ? 0xa000 // VRAM
+	     : address < 0xc000 ? 0xc000 // SRAM
+	     : address < 0xd000 ? 0xd000 // WRAM
+	     : 0xffff; // HRAM
+}
+
+void symbol_append(struct Symbol **symbols, const char *name, int bank, int address) {
+	size_t name_len = strlen(name) + 1;
+	struct Symbol *symbol = xmalloc(sizeof(*symbol) + name_len);
+	symbol->address = address;
+	symbol->offset = bank > 0 ? address + (bank - 1) * get_address_type_limit(address) : address;
```

Comment:
I think this quickly breaks down with WRAM addresses. While it may not be used for those, I'd just make `get_address_type_limit` return the size of the region instead of its limit. That or just symbol->offset == symbol->address for anything that isn't ROMX.